lib/mayaUsd/fileio/shading/shadingModeUseRegistry.cpp

```diff
@@ -119,35 +120,45 @@ class UseRegistryShadingModeExporter : public UsdMayaShadingModeExporter
             const SdfPath shaderUsdPath =
                 parentPath.AppendChild(shaderUsdPrimName);
 
-            UsdMayaPrimWriterSharedPtr primWriter =
-                context.GetWriteJobContext().CreatePrimWriter(
-                    depNodeFn,
-                    shaderUsdPath);
+            UsdMayaShaderWriterRegistry::WriterFactoryFn primWriterFactory
+                = UsdMayaShaderWriterRegistry::Find(
+                    TfToken(depNodeFn.typeName().asChar()), context.GetExportArgs());
+            if (!primWriterFactory) {
+                return nullptr;
+            }
 
-            UsdMayaShaderWriterSharedPtr shaderWriter =
-                std::dynamic_pointer_cast<UsdMayaShaderWriter>(primWriter);
+            UsdMayaPrimWriterSharedPtr primWriter
+                = primWriterFactory(depNodeFn, shaderUsdPath, context.GetWriteJobContext());
+            if (!primWriter) {
+                return nullptr;
+            }
+
+            UsdMayaShaderWriterSharedPtr shaderWriter
+                = std::dynamic_pointer_cast<UsdMayaShaderWriter>(primWriter);
 
             // Store the shader writer pointer whether we succeeded or not so
             // that we don't repeatedly attempt and fail to create it for the
             // same node.
             shaderWriterMap[nodeHandle] = shaderWriter;
 
+            shaderWriter->Write(UsdTimeCode::Default());
```

Comment:
Make sure the shaderWriter writes only once. When returning an existing writer from the cache above, it will not write a second time.

Reply:
I don't think we should be doing **any** authoring here. That seems like an odd side effect for a function that sounds like it just gets you the thing you need to do the authoring.